File: backend/app/models/student.py
from datetime import datetime, date

from sqlalchemy import String, Date, Integer, Boolean, DateTime, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Student(Base):
    __tablename__ = "students"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(50), nullable=False)
    gender: Mapped[str] = mapped_column(String(10), nullable=False)
    birth_date: Mapped[date] = mapped_column(Date, nullable=False)
    group_name: Mapped[str] = mapped_column(String(50), default="")
    enrollment_year: Mapped[int] = mapped_column(Integer, nullable=False)
    parent_id: Mapped[int] = mapped_column(ForeignKey("parents.id"), nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="active")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

    parent: Mapped["Parent"] = relationship(back_populates="students")
    coach_associations: Mapped[list["CoachStudent"]] = relationship(back_populates="student")


class CoachStudent(Base):
    __tablename__ = "coach_students"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    coach_id: Mapped[int] = mapped_column(ForeignKey("coaches.id"), nullable=False)
    student_id: Mapped[int] = mapped_column(ForeignKey("students.id"), nullable=False)
    year: Mapped[int] = mapped_column(Integer, nullable=False)
    is_primary: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    coach: Mapped["Coach"] = relationship()
    student: Mapped["Student"] = relationship(back_populates="coach_associations")

File: frontend/src/api/student.ts
import http from './index'

export function getStudentsApi(params?: { keyword?: string; group?: string }) {
  return http.get('/students', { params })
}

export function getStudentApi(id: number) {
  return http.get(`/students/${id}`)
}

export function createStudentApi(data: {
  name: string
  gender: string
  birth_date: string
  group_name: string
  enrollment_year: number
  parent_id?: number
}) {
  return http.post('/students', data)
}

export function updateStudentApi(
  id: number,
  data: {
    name?: string
    gender?: string
    birth_date?: string
    group_name?: string
    enrollment_year?: number
    parent_id?: number
  }
) {
  return http.put(`/students/${id}`, data)
}
